File: python/mestra/evaluate.py
"""Evaluating a file's callables on a keys table.

Section 10: evaluating a file on a keys table yields a file with the
same slots, now holding data. Distillation is that operation on a
grid.

What the result carries under `/callables`. Section 10 says only
that the slots now hold data, and the four implementations read that
sentence three different ways, which is finding 7 of the Phase 3
report: two leave no `/callables` group, one leaves it present and
empty, one leaves the whole callable with its dictionary. All three
conform -- section 13 lets a container group be absent or present
and empty, and no rule forbids a callable nothing references -- so
the sentence alone does not settle it and another one has to.

Section 12 is that sentence: "A distilled lookup table is stored
data and therefore public; the callable that produced it, its
history, and its validation records are not." The evaluated file is
the distilled table. The callable is the thing section 12 says is
not part of it, so this implementation leaves no `/callables` group
at all: the result is a plain data file with the same keys, the same
supports and the same slot attributes, and every slot's `source` is
`data`.

Section 13's rule that "a writer reproducing a file it read keeps
whichever of the two it found" is about a round trip and not about
this: evaluation does not reproduce the file it read, it produces a
different file, and a group that would be empty may be absent.

Nothing here reads `/private` or copies it. An evaluated file is a
new file and not a round trip of the one the callables came from;
`read` and `write` are what preserve a producer's own records.
"""

from __future__ import annotations

import copy
from collections.abc import Mapping
from typing import Any

import numpy as np

from .callables import keys_table, table_length
from .errors import MestraError
from .model import (
    ArraySlot,
    Dataset,
    Key,
    MemorySource,
    ScalarSlot,
    Storage,
)

__all__ = ["evaluate"]


def evaluate(dataset: Dataset, table: Any,
             names: list[str] | None = None) -> Dataset:
    """Evaluate every callable slot on a keys table.

    `table` is a keys table as section 26 defines it for Python: a
    mapping from key name to a one-dimensional array, or a
    two-dimensional array of shape (rows, keys) together with
    `names` in the file's key order.

    The result is a new dataset whose rows are the table's rows and
    whose callable slots hold the values the callables produced.
    Every slot's `source` is `data`, and the result carries no
    callables and no `/private`: section 12 says the distilled table
    is public and that the callable that produced it, its history
    and its validation records are not. The module docstring has the
    argument in full.
    """
    columns = keys_table(table, names)
    rows = table_length(columns)
    missing = [n for n in dataset.key_names() if n not in columns]
    if missing:
        raise MestraError(
            "", "the keys table needs one column per key the file "
            "declares, in the file's key order; it has none for %s"
            % ", ".join(missing))

    out = Dataset(writer=dataset.writer, created=dataset.created,
                  format=dataset.format,
                  generalisation_group=dataset.generalisation_group,
                  notes=dataset.notes)
    out.n_rows = rows
    out.categories = copy.deepcopy(dataset.categories)
    out.extra = dict(dataset.extra)
    out.present = {name for name in dataset.present
                   if name != "callables"}

    for name in dataset.key_names():
        source = dataset.keys[name]
        out.keys[name] = Key(
            name, source.role, units=source.units, lower=source.lower,
            upper=source.upper, category=source.category,
            trajectory_group=source.trajectory_group,
            parent=source.parent, itemsize=source.itemsize,
            data=MemorySource(_column(columns[name], source)),
            extra=dict(source.extra))

    produced = _produce(dataset, columns)

    for name, slot in dataset.scalars.items():
        out.scalars[name] = _fill(slot, produced, rows, ScalarSlot)
    for name, support in dataset.supports.items():
        copied = copy.copy(support)
        copied.dataset = out
        copied.node_arrays = {}
        copied.cell_arrays = {}
        # The result stands on its own, so its arrays are in memory
        # and not still in the file the callables came from.
        copied._cells = {
            which: (None if source is None
                    else MemorySource(source.read()))
            for which, source in support._cells.items()}
        if support.coordinates is not None:
            copied.coordinates = _fill(support.coordinates, produced,
                                       rows, ArraySlot, copied)
        for where, arrays in (("node_arrays", support.node_arrays),
                              ("cell_arrays", support.cell_arrays)):
            into = getattr(copied, where)
            for slot_name, array in arrays.items():
                into[slot_name] = _fill(array, produced, rows,
                                        ArraySlot, copied)
        out.supports[name] = copied
    if dataset.has_row_support:
        out.row_support = dataset.row_support
    return out


def _column(values: Any, key: Key) -> np.ndarray:
    """A table column in the dtype the key's role requires."""
    array = np.asarray(values)
    if key.role in ("categorical", "group", "split", "status"):
        return array.astype("<i4", copy=False)
    if key.role == "id":
        if array.dtype.kind in "US":
            return array
        return array.astype("<i8", copy=False)
    return array.astype("<f8", copy=False)


def _produce(dataset: Dataset,
             columns: Mapping[str, np.ndarray]) -> dict[str, Any]:
    """Call each callable once, whatever number of slots it fills."""
    wanted = set()
    for slot in dataset.slots().values():
        if slot.is_callable:
            identifier = slot.callable_id or ""
            if identifier not in dataset.callables:
                raise MestraError(
                    "E14", "this slot names the callable %r and the "
                    "file does not hold it" % identifier, slot.name)
            wanted.add(identifier)
    return {name: dataset.callables[name](columns) for name in wanted}


def _fill(slot: Any, produced: Mapping[str, Any], rows: int,
          kind: Any, support: Any = None) -> Any:
    """One slot of the evaluated dataset."""
    made = copy.copy(slot)
    made.storage = Storage()
    made.extra = dict(slot.extra)
    if support is not None:
        made.support = support
    if not slot.is_callable:
        if slot.data is not None and "row" in slot.dims:
            have = int(slot.data.shape[0])
            if have != rows:
                raise MestraError(
                    "E16", "this slot holds %d rows of stored data and "
                    "the keys table has %d; evaluate a file whose "
                    "row-varying slots are all callables, or pass a "
                    "table of the same length" % (have, rows),
                    slot.name)
            made.data = MemorySource(slot.data.read())
        elif slot.data is not None:
            made.data = MemorySource(slot.data.read())
        return made
    values = produced[slot.callable_id or ""]
    output = slot.output or slot.name
    if output not in values:
        raise MestraError(
            "E14", "the callable %s produced no output called %r"
            % (slot.callable_id, output), slot.name)
    array = np.asarray(values[output], dtype="<f8")
    _check_shape(slot, array, rows, kind)
    made.data = MemorySource(array)
    made.source = "data"
    made.output = None
    return made


def _check_shape(slot: Any, array: np.ndarray, rows: int,
                 kind: Any) -> None:
    if array.shape[:1] != (rows,):
        raise MestraError(
            "E16", "a callable returns one entry per table row; this "
            "one returned %s for %d rows" % (array.shape, rows),
            slot.name)
    if kind is ScalarSlot:
        if array.ndim != 1:
            raise MestraError(
                "E04", "a scalar slot takes one value per row; the "
                "callable returned %s" % (array.shape,), slot.name)
        return
    if array.ndim != len(slot.dims):
        raise MestraError(
            "E04", "this slot has the dimensions %s; the callable "
            "returned %s" % (", ".join(slot.dims), array.shape),
            slot.name)
    if array.shape[-1] != slot.components:
        raise MestraError(
            "E31", "this slot declares %d components and the callable "
            "returned %d" % (slot.components, array.shape[-1]),
            slot.name)
    support = slot.support
    if support is not None:
        width = (support.n_nodes if slot.location == "node"
                 else support.n_cells)
        if array.shape[-2] != width:
            raise MestraError(
                "E05", "this support has %d %ss and the callable "
                "returned %d" % (width, slot.location, array.shape[-2]),
                slot.name)

File: matlab/+mestra/evaluate.m
function out = evaluate(dataset, keysTable)
%MESTRA.EVALUATE  Materialise a dataset's callable slots on a keys table.
%
%   OUT = MESTRA.EVALUATE(DATASET, KEYSTABLE) returns a new
%   mestra.Dataset with the same slots as DATASET, every slot that was
%   served by a callable now holding data, and the key columns holding
%   the rows of KEYSTABLE.  Specification section 10 calls this
%   evaluating the file on a keys table; doing it on a grid is
%   distillation.
%
%   KEYSTABLE is a MATLAB table whose variable names are the key names
%   (section 26).  Its row order is the evaluation order: output row i
%   is the result for table row i.  Every key the file declares must
%   have a column; other columns are passed through to the callables,
%   which read the columns they declare and ignore the rest.
%
%   DATASET is not changed.  A slot that already held data is carried
%   over unchanged when it does not vary along `row`; when it does,
%   its row count must match the table, and an error names the slot if
%   it does not.
%
%   Example
%
%       d = mestra.read('affine_zero_rows.mes');
%       t = table(0.5, 4.0, 'VariableNames', {'mach', 'alpha'});
%       e = mestra.evaluate(d, t);
%       e.scalar('cl').values                    % 1.45
%       a = e.nodeArray('s0', 'pressure');
%       mestra.permute(a.values, a.dims, {'row', 'node'})
%
%   See also mestra.Callable, mestra.Affine, mestra.read, mestra.write.

    if ~isa(dataset, 'mestra.Dataset')
        error('mestra:evaluate', ...
              'the first argument must be a mestra.Dataset');
    end
    if ~istable(keysTable)
        error('mestra:keysTable', ...
              'the keys table must be a MATLAB table (section 26)');
    end
    nRows = height(keysTable);
    declared = dataset.keyNames();
    missing = setdiff(declared, keysTable.Properties.VariableNames);
    if ~isempty(missing)
        error('mestra:keysTable', ...
              'the keys table has no column for %s', strjoin(missing, ', '));
    end

    out = copyDataset(dataset);
    out.path = '';
    out.nRows = nRows;
    for i = 1:numel(out.keys)
        name = out.keys(i).name;
        column = keysTable.(name);
        if iscell(column) || isstring(column)
            out.keys(i).values = reshape(cellstr(column), 1, []);
        else
            out.keys(i).values = reshape(column, 1, []);
        end
    end
    if ~isempty(out.rowSupport) && numel(out.rowSupport) ~= nRows
        out.rowSupport = [];
    end

    results = containers.Map('KeyType', 'char', 'ValueType', 'any');
    for i = 1:numel(out.scalars)
        out.scalars(i) = fill(out, results, out.scalars(i), ...
                              ['/scalars/' out.scalars(i).name], ...
                              keysTable, nRows, {'row'});
    end
    for i = 1:numel(out.supports)
        s = out.supports(i);
        base = ['/supports/' s.name];
        if ~isempty(s.coordinates)
            s.coordinates = fill(out, results, s.coordinates, ...
                [base '/coordinates'], keysTable, nRows, ...
                {'row', 'node', 'component'});
        end
        for j = 1:numel(s.nodeArrays)
            s.nodeArrays(j) = fill(out, results, s.nodeArrays(j), ...
                [base '/node_arrays/' s.nodeArrays(j).name], keysTable, ...
                nRows, {'row', 'node', 'component'});
        end
        for j = 1:numel(s.cellArrays)
            s.cellArrays(j) = fill(out, results, s.cellArrays(j), ...
                [base '/cell_arrays/' s.cellArrays(j).name], keysTable, ...
                nRows, {'row', 'cell', 'component'});
        end
        out.supports(i) = s;
    end
    out.callables = mestra.Dataset.emptyCallable();
end

% ---------------------------------------------------------------------

function slot = fill(d, results, slot, path, keysTable, nRows, fileDims)
%fill  One slot: call its callable, or check the data it already has.
    if strcmp(slot.source, 'data')
        if strcmp(slot.varies, 'row') && ~isempty(slot.values)
            n = size(slot.values, numel(slot.dims));
            if n ~= nRows
                error('mestra:evaluate', ...
                      ['the slot "%s" holds %d rows of data and the keys ' ...
                       'table has %d; drop it or pass a matching table'], ...
                      path, n, nRows);
            end
        end
        return
    end
    [id, output] = splitSource(slot.source, slot.output, path);
    if ~results.isKey(id)
        record = d.callable(id);
        if isempty(record.obj)
            error('mestra:evaluate', ...
                  ['the callable "%s" has type "%s", which no registered ' ...
                   'type can evaluate; register it with ' ...
                   'mestra.Registry.register'], id, record.type);
        end
        results(id) = record.obj.call(keysTable);
    end
    produced = results(id);
    if ~produced.isKey(output)
        error('mestra:evaluate', ...
              'the callable "%s" produces no output named "%s"', id, output);
    end
    value = produced(output);
    slot.source = 'data';
    slot.output = '';
    slot.dtype = 'float64';
    if numel(fileDims) == 1
        slot.values = reshape(double(value.buffer()), 1, []);
        slot.dims = {'row'};
    else
        buf = double(value.buffer());
        slot.values = reshape(buf, [fliplr(value.shape) 1 1]);
        slot.dims = fliplr(fileDims);
        if isempty(slot.components)
            slot.components = size(slot.values, 1);
        end
    end
    if isfield(slot, 'shape')
        sz = size(slot.values);
        sz = [sz ones(1, numel(slot.dims) - numel(sz))];
        slot.shape = fliplr(sz(1:numel(slot.dims)));
    end
end

function [id, output] = splitSource(source, output, path)
    if numel(source) <= 9 || ~strncmp(source, 'callable:', 9)
        error('mestra:E36', ...
              'the slot "%s" has source "%s"', path, source);
    end
    id = source(10:end);
    if isempty(output)
        error('mestra:E39', ...
              'the slot "%s" names a callable and no output', path);
    end
end

function out = copyDataset(d)
%copyDataset  A fresh Dataset with the same contents.
    out = mestra.Dataset();
    out.format = d.format;
    out.writer = d.writer;
    out.created = d.created;
    out.aligned = d.aligned;
    out.generalisationGroup = d.generalisationGroup;
    out.nRows = d.nRows;
    out.keys = d.keys;
    out.scalars = d.scalars;
    out.categories = d.categories;
    out.supports = d.supports;
    out.callables = d.callables;
    out.rowSupport = d.rowSupport;
    out.notes = d.notes;
    out.privateTree = d.privateTree;
    out.unknownGroups = d.unknownGroups;
    out.skipped = d.skipped;
    if d.unknownAttrs.Count > 0
        out.unknownAttrs = containers.Map(d.unknownAttrs.keys(), ...
                                          d.unknownAttrs.values());
    end
    out.groupsPresent = d.groupsPresent;
end
